### PR TITLE
Initial version of the WebSocket Next documentation

### LINKED ISSUE #39144: WebSockets Next: documentation
### Description

Follow-up of https://github.com/quarkusio/quarkus/pull/39142.

### Implementation ideas

_No response_

### PR DESCRIPTION
**Pull request by cescoffier**
Fix https://github.com/quarkusio/quarkus/issues/39144.

### CODE REVIEW

**Comment by cescoffier** [comment]
This is the very initial version. I tried to cover the current implementation. Let me know if I missed anything.

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+
+WebSocket is a low-level transport protocol, in contrast to HTTP.
+Message formats, routing, or processing require prior agreement between the client and server regarding message semantics.
+
+For WebSocket clients and servers, the `Sec-WebSocket-Protocol` header in the HTTP handshake request allows negotiation of a higher-level messaging protocol. In its absence, the server and client must establish their own conventions.
+
+== Quarkus WebSockets vs. Quarkus WebSockets Next
+
+This guide utilizes the `quarkus-websockets-next` extension, an implementation of the WebSocket API boasting enhanced efficiency and usability compared to the legacy `quarkus-websockets` extension.
+The original `quarkus-websockets` extension remains accessible and will receive ongoing support.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:44]:
```suggestion
The original `quarkus-websockets` extension remains accessible, will receive ongoing support, but it's unlikely to receive to feature development.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+
+
+=== WebSocket endpoint methods
+
+A WebSocket endpoint comprises the following components:
+
+* Path: This is the URL path where the WebSocket connection is established (e.g., ws://localhost:8080/).
+* Maximum one `@OnTextMessage` method: Handles the connected client's text messages.
+* Maximum one `@OnBinaryMessage` method: Handles the binary messages the connected client sends.
+* Maximum one `@OnOpen` method: Invoked when a client connects to the WebSocket.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:192]:
```suggestion
* At most one `@OnOpen` method: Invoked when a client connects to the WebSocket.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+* Maximum one `@OnClose` method: Executed upon the client disconnecting from the WebSocket.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:193]:
```suggestion
* At most one `@OnClose` method: Executed upon the client disconnecting from the WebSocket.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+
+Only some endpoints need to include all methods.
+However, it must contain at least `@On[Text|Binary]Message` or `@OnOpen`.
+
+An error is thrown at build time if any endpoint violates these rules.
+The inner classes representing sub-websockets adhere to the same guidelines.
+
+IMPORTANT: Any methods annotated with `@OnTextMessage`, `@OnBinaryMessage`, `@OnOpen`, and `@OnClose` outside a WebSocket endpoint are considered erroneous and are flagged during the build process.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:201]:
```suggestion
IMPORTANT: Any methods annotated with `@OnTextMessage`, `@OnBinaryMessage`, `@OnOpen`, and `@OnClose` outside a WebSocket endpoint are considered erroneous and will result in the build failing with an appropriate error message.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+
+`OnTextMessage` are invoked for every _text_ message received from the client.
+`OnBinaryMessage` are invoked for every _binary_ message the client receives.
+
+=== Invocation Rules
+
+When invoking these annotated methods, the _session_ scope linked to the WebSocket connection remains active.
+In addition, the request scope is active until the completion of the method (or until it produces its result for async and reactive methods).
+
+Quarkus WebSocket Next supports _blocking_ and _non-blocking_ logic, akin to RESTEasy Reactive, determined by the method signature and additional annotations such as `@Blocking` and `@NonBlocking`.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:215]:
```suggestion
Quarkus WebSocket Next supports _blocking_ and _non-blocking_ logic, akin to Quarkus REST, determined by the method signature and additional annotations such as `@Blocking` and `@NonBlocking`.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+* `Uni<X>`: Specifies a non-blocking method where the item emitted by the non-null `Uni` is sent to the client as a response.
+* `Multi<X>`: Indicates a non-blocking method where the items emitted by the non-null `Multi` are sequentially sent to the client until completion or cancellation.
+
+Here are some examples of these methods:
+
+[source, java]
+----
+@OnTextMessage
+void consume(Message m) {
+// Process the incoming message. The method is called for each incoming message.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:257]:
```suggestion
// Process the incoming message. The method is called on an executor thread for each incoming message.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+}
+
+@OnTextMessage
+Uni<Void> consumeAsync(Message m) {
+// Process the incoming message. The method is called for each incoming message.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:262]:
```suggestion
// Process the incoming message. The method is called on an event loop thread for each incoming message.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 8]
```diff
@@ -0,0 +1,546 @@
...
+connection.sendTextAndAwait("Hello!");
+
+// Broadcast messages:
+connection.broadcast().sendTextAndAwait(departure);
+
+// Access path parameters:
+String param = connection.pathParam("foo");
+----
+
+The `WebSocketConnection` provides both blocking and non-blocking method to send messages:
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:454]:
```suggestion
The `WebSocketConnection` provides both a blocking and a non-blocking method to send messages:
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+
+- `sendTextAndAwait(String message)`: Sends a text message to the client and waits for the message to be sent. It's blocking.
```
Comment (geoand) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:456]:
```suggestion
- `sendTextAndAwait(String message)`: Sends a text message to the client and waits for the message to be sent. It's blocking and should only be called from an executor thread.
```

**Review by geoand — COMMENTED** [review]
Very nice!

I started off with some comments

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+----
+@Inject io.quarkus.websockets.next.WebSocketConnection session;
+// ...
+String value = session.pathParam("parameterName");
+----
+
+Path parameter values are always strings.
+If the path parameter is not present in the path, the `pathParam` method returns `null`.
+
+NOTE: Query parameters are not supported.
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:122]: However, you can access the query string of the handshake request with `session.handshakeRequest().query()`.

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+Client interaction occurs through emitting HTTP requests with appropriate methods and paths, following a request-response pattern.
+The server routes incoming requests to corresponding handlers based on path, method, and headers and then replies with a well-defined response.
+
+Conversely, WebSocket typically involves a single endpoint for the initial HTTP connection, after which all messages utilize the same TCP connection.
+It introduces an entirely different interaction model: asynchronous and message-driven.
+
+WebSocket is a low-level transport protocol, in contrast to HTTP.
+Message formats, routing, or processing require prior agreement between the client and server regarding message semantics.
+
+For WebSocket clients and servers, the `Sec-WebSocket-Protocol` header in the HTTP handshake request allows negotiation of a higher-level messaging protocol. In its absence, the server and client must establish their own conventions.
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:39]: This reminds me that we might add something convenient here as well. Currently, the only thing you can do is obtain the header value with `connection.handshakeRequest().header("Sec-WebSocket-Protocol")`.

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+----
+
+Path parameter values are always strings.
+If the path parameter is not present in the path, the `pathParam` method returns `null`.
+
+NOTE: Query parameters are not supported.
+
+=== Sub-websockets endpoints
+
+A class annotated with `@WebSocket` can encapsulate (static) inner classes, which are also annotated with `@WebSocket` and represent _sub-web_ sockets.
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:126]: I believe that the correct terminology is "static nested classes" ;-)

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+When the `@OnOpen` method is invoked, a session scope corresponding to the WebSocket connection is established.
+Subsequent calls to `@On[Text|Binary]Message` or `@OnClose` methods utilize this same session scope.
+The session scope remains active until the `@OnClose` method completes execution, at which point it is terminated.
+
+The `WebSocketConnection` object, which represents the connection itself, is also a session-scoped bean, allowing developers to access and manage WebSocket-specific data within the context of the session.
+
+In cases where a WebSocket endpoint does not declare an `@OnOpen` method, the session scope is still created.
+It remains active until the connection terminates, regardless of the presence of an `@OnClose` method.
+
+Methods annotated with `@OnTextMessage,` `@OnBinaryMessage,` `@OnOpen`, and `@OnClose` haves also have the request scoped activated for the duration of the method execution (until it produced its result).
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:182]:
```suggestion
Methods annotated with `@OnTextMessage,` `@OnBinaryMessage,` `@OnOpen`, and `@OnClose` also have the request scoped activated for the duration of the method execution (until it produced its result).
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+* At most one `@OnTextMessage` method: Handles the connected client's text messages.
+* At most one `@OnBinaryMessage` method: Handles the binary messages the connected client sends.
+* At most one `@OnOpen` method: Invoked when a client connects to the WebSocket.
+* At most one `@OnClose` method: Executed upon the client disconnecting from the WebSocket.
+
+Only some endpoints need to include all methods.
+However, it must contain at least `@On[Text|Binary]Message` or `@OnOpen`.
+
+An error is thrown at build time if any endpoint violates these rules.
+The inner classes representing sub-websockets adhere to the same guidelines.
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:199]:
```suggestion
The static nested classes representing sub-websockets adhere to the same guidelines.
```

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+These methods can accept parameters in two formats:
+
+* The message object (of any type).
+* A `Multi<X>` with X as the message type.
+* Any other parameters should be flagged as errors.
+
+The message object represents the data sent and can be accessed as either raw content (`String`, `JsonObject`, `JsonArray`, `Buffer` or `byte[]`) or deserialized high-level objects, which is the recommended approach.
+
+When receiving a `Multi`, the method is invoked once per connection, and the provided `Multi` receives the items transmitted by this connection.
+The method must subscribe to the `Multi` to receive these items (or return a Multi).
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:238]: TBH I find the `Multi` param use case quite confusing. First, the method is not executed when a message is received but immediately after the connection is open and second you must not forget to subscribe to the `Multi` (unless returning it).

Maybe we should introduce a separate annotation to consume messages using `Multi`? Something like `@TextMessageStream`?

File: docs/src/main/asciidoc/websockets-next-reference.adoc [hunk 16]
```diff
@@ -0,0 +1,546 @@
...
+////
+= WebSockets Next Extension Reference Guide
+:extension-status: preview
+include::_attributes.adoc[]
+:numbered:
+:sectnums:
+:categories: web
+:topics: web,websockets
+:extensions: io.quarkus:quarkus-websockets-next
+
```
Comment (mkouba) [inline docs/src/main/asciidoc/websockets-next-reference.adoc:14]: I think that both guides should start with a section like "Exprimental status"  that would explain that the API may change in the future. WDYT?

**Comment by mkouba** [comment]
Thanks Clement! I've added some comments...

File: docs/src/main/asciidoc/websockets-next-reference.adoc (hunk 11 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/websockets-next-reference.adoc:39]: Can you open an issue so we do not forget?

File: docs/src/main/asciidoc/websockets-next-reference.adoc (hunk 15 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/websockets-next-reference.adoc:238]: I'm not sure we need to introduce a new annotation. Reactive messaging typically reuses the same. Same for the rest client.

Comment (cescoffier) [reply docs/src/main/asciidoc/websockets-next-reference.adoc:238]: We can see if the users are confused.

File: docs/src/main/asciidoc/websockets-next-reference.adoc (hunk 16 above)
Comment (cescoffier) [reply docs/src/main/asciidoc/websockets-next-reference.adoc:14]: Done.

**Comment by cescoffier** [comment]
Should we merge it? It would be great to have documentation at the extension will be released soon.

**Comment by geoand** [comment]
+1

**Comment by mkouba** [comment]
@cescoffier is there anything else you would like to add? If not I think that we're ready to merge this pull request.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I'm going to release 3.9.0 core artifacts so I will merge this to make sure we have it in 3.9.

**Pull request merged**